### PR TITLE
Mono Kotlin extensions and functions fixes

### PR DESCRIPTION
**Pull request by sdeleuze**
- Polishing
- Better consistency in term of file naming and functions dispatching
- Make `whenComplete` alias working on `Publisher<*>` not just `Publisher<Void>`

### CODE REVIEW

File: reactor-core/src/test/kotlin/reactor/core/publisher/MonoFunctionsTests.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2011-2017 Pivotal Software Inc, All Rights Reserved.
```
Comment (bsideup) [inline reactor-core/src/test/kotlin/reactor/core/publisher/MonoFunctionsTests.kt]:
```suggestion
  * Copyright (c) 2011-2018 Pivotal Software Inc, All Rights Reserved.
```

File: reactor-core/src/test/kotlin/reactor/core/publisher/MonoFunctionsTests.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    @Test
+    fun `whenComplete with two Monos`() {
+        StepVerifier.create(whenComplete("foo1".toMono(), "foo2".toMono()))
+                .verifyComplete()
+    }
+
+    @Test
+    fun `zip with Monos and combinator`() {
+        StepVerifier.create(zip("foo1".toMono(), "foo2".toMono(), "foo3".toMono()) { it.joinToString() })
```
Comment (bsideup) [inline reactor-core/src/test/kotlin/reactor/core/publisher/MonoFunctionsTests.kt]: could you please extract `zip()`? Block outside of the zip's parentheses makes it a bit hard to read (tbh I was like, why `.create` has a block as the last argument? 😅

File: reactor-core/src/test/kotlin/reactor/core/publisher/MonoExtensionsTests.kt [hunk 3]
```diff
@@ -190,7 +191,7 @@ class MonoExtensionsTests {
     fun onErrorResume() {
         StepVerifier.create(IOException()
                 .toMono<String>()
-                .onErrorResume(IOException::class, { "foo".toMono() }))
+                .onErrorResume(IOException::class) { "foo".toMono() })
```
Comment (bsideup) [inline reactor-core/src/test/kotlin/reactor/core/publisher/MonoExtensionsTests.kt]: Same issue as in `MonoFunctionsTests` - reads as `(IOException::class) { "foo".toMono() }` 😊

File: reactor-core/src/main/kotlin/reactor/core/publisher/MonoFunctions.kt [hunk 4]
```diff
@@ -17,36 +17,17 @@
...
 /**
  * Aggregates the given void [Publisher]s into a new void [Mono].
  * An alias for a corresponding [Mono.when] to avoid use of `when`, which is a keyword in Kotlin.
  *
  * @author DoHyung Kim
+ * @author Sebastien Deleuze
  * @since 3.1
  */
-fun whenComplete(vararg sources: Publisher<Void>): Mono<Void> = MonoBridges.`when`(sources)
+fun whenComplete(vararg sources: Publisher<*>): Mono<Void> = MonoBridges.`when`(sources)
```
Comment (bsideup) [inline reactor-core/src/main/kotlin/reactor/core/publisher/MonoFunctions.kt]: doesn't it break the public binary API? Maybe we should add an overloaded method instead?

File: reactor-core/src/main/kotlin/reactor/core/publisher/MonoFunctions.kt [hunk 5]
```diff
@@ -17,36 +17,17 @@
 package reactor.core.publisher
```
Comment (bsideup) [inline reactor-core/src/main/kotlin/reactor/core/publisher/MonoFunctions.kt]: is it safe to rename extension classes in Kotlin?

File: reactor-core/src/main/kotlin/reactor/core/publisher/MonoFunctions.kt (hunk 4 above)
Comment (sdeleuze) [reply reactor-core/src/main/kotlin/reactor/core/publisher/MonoFunctions.kt]: I have asked on Kotlin Slack to be sure.

File: reactor-core/src/main/kotlin/reactor/core/publisher/MonoFunctions.kt (hunk 5 above)
Comment (sdeleuze) [reply reactor-core/src/main/kotlin/reactor/core/publisher/MonoFunctions.kt]: I have asked on Kotlin Slack to be sure.

File: reactor-core/src/main/kotlin/reactor/core/publisher/MonoFunctions.kt (hunk 4 above)
Comment (sdeleuze) [reply reactor-core/src/main/kotlin/reactor/core/publisher/MonoFunctions.kt]: Generic type will be erased so this is not a breaking change.

**Comment by sdeleuze** [comment]
@bsideup I think I have taken in account all your remarks.

**Comment by sdeleuze** [comment]
I have made other changes to preserve backward compatibility with `@file:JvmName("MonoWhenFunctionsKt")` + moving back extensions where they were. We can't deprecate and put them in the right place because there is a clash in signature, so I suggest doing that for now and fixing the TODO I put in the comments in Reactor Core 3.3.

File: reactor-core/src/test/kotlin/reactor/core/publisher/FluxExtensionsTests.kt [hunk 6]
```diff
@@ -162,7 +162,7 @@ class FluxExtensionsTests {
         StepVerifier
                 .create(IOException()
                         .toFlux<String>()
-                        .onErrorResume(IOException::class, { "foo".toMono() }))
+                        .onErrorResume(IOException::class) { "foo".toMono() })
```
Comment (bsideup) [inline reactor-core/src/test/kotlin/reactor/core/publisher/FluxExtensionsTests.kt]: could you please improve the readability of this one as well? (as you did [here](https://github.com/reactor/reactor-core/pull/1475/commits/3f29a64a72ed969bad9a703aff8ddc33a33e60af))

Comment (simonbasle) [reply reactor-core/src/test/kotlin/reactor/core/publisher/FluxExtensionsTests.kt]: (I did the update)

**Pull request merged**